plaidml/bridge/openvino/ops/embedding_segments_sum.cpp

```diff
@@ -0,0 +1,101 @@
+// Copyright (C) 2020 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "plaidml_ops.hpp"
+#include "plaidml_util.hpp"
+
+#include "ngraph/opsets/opset.hpp"
+#include "ngraph/opsets/opset4.hpp"
+
+#include "plaidml/op/op.h"
+
+using namespace plaidml;          // NOLINT[build/namespaces]
+using namespace InferenceEngine;  // NOLINT[build/namespaces]
+
+namespace {
+
+template <typename T>
+std::vector<T> cast_constant_operand(size_t operand_idx, ngraph::Node* layer) {
+  auto* ngraph_const = ngraph::as_type<ngraph::op::Constant>(layer->get_input_node_ptr(operand_idx));
+  if (ngraph_const) {
+    return ngraph_const->cast_vector<T>();
+  } else {
+    THROW_IE_EXCEPTION << "Dynamic slicing not currently supported by PlaidML plugin; all of indices, segment_ids, "
+                          "num_segments and default index must be Constants.";
+  }
+}
+
+}  // namespace
+
+namespace PlaidMLPlugin {
+
+void registerEmbeddingSegmentsSum() {
+  registerOp("EmbeddingSegmentsSum", [](const Context& ctx) {
+    auto* layer = ngraph::as_type<ngraph::opset4::EmbeddingSegmentsSum>(ctx.layer);
+    IE_ASSERT(ctx.operands.size() >= 4);
+    IE_ASSERT(ctx.operands.size() <= 6);
+    auto I = ctx.operands.at(0);
+    auto indices = ctx.operands.at(1);
+    IE_ASSERT(indices.rank() == 1);
+    auto segment_ids = ctx.operands.at(2);
+    auto num_segments = cast_constant_operand<int32_t>(3, layer)[0];
+    int32_t default_index = -1;
+    edsl::Tensor per_sample_weights;
+    bool with_weights = false;
+
+    if (ctx.operands.size() >= 5) {
+      default_index = cast_constant_operand<int32_t>(4, layer)[0];
+      if (ctx.operands.size() == 6) {
+        per_sample_weights = ctx.operands.at(5);
+        with_weights = true;
+      }
+    }
+
+    edsl::Tensor I_gathered = edsl::gather(I, indices);
+
+    auto ndims = I_gathered.rank();
+    std::vector<edsl::TensorDim> I_dims(ndims);
+    std::vector<edsl::TensorIndex> I_idxs(ndims);
+    I_gathered.bind_dims(I_dims);
+    auto O_dims = I_dims;
+    auto O_idxs = I_idxs;
+
+    // Create zero-initialized input tensor.
+    O_dims[0] = edsl::TensorDim(num_segments);
+    edsl::Tensor scatter_in_shape = edsl::Contraction(O_dims, O_idxs).assign(I(I_idxs)).build();
+    auto zero = cast(edsl::Tensor{0}, I_gathered.dtype());
+    auto scatter_shape_vec = scatter_in_shape.compute_shape().sizes();
+    std::vector<int> scatter_shape(begin(scatter_shape_vec), end(scatter_shape_vec));
+    std::vector<int> target_axes = {};
+    auto I_scatter_init = op::broadcast(zero, scatter_shape, target_axes);
```

Comment:
I wonder if we need to extend broadcast to accept a target tensor instead of a target shape, so we can avoid using compute_shape(). But for now I think it is ok in this PR.

Reply:
Yes I think `broadcast` API is so limited and we could propose this to plaidml team.

Reply:
By the way. I remember Tim & Frank mentioned that a `like` operation should be added so this `broadcast` will be replaced when that is done.

Reply:
Thanks, we should keep this in mind.